elftools/ehabi/ehabiinfo.py

```diff
@@ -0,0 +1,217 @@
+                Return None when corrupt or unwindable.
+
+            eh_table_offset:
+                Integer.
+                Return None when corrupt or unwindable or ARM inline compact.
+
+            unwindable:
+                bool. Whether this function is unwindable.
+
+            corrupt:
+                bool. Whether this entry is corrupt.
+
+    """
+
+    def __init__(self,
+                 function_offset,
+                 personality,
+                 bytecode_array,
+                 eh_table_offset=None,
+                 unwindable=True,
+                 corrupt=False
+                 ):
+        self.function_offset = function_offset
+        self.personality = personality
+        self.bytecode_array = bytecode_array
+        self.eh_table_offset = eh_table_offset
+        self.unwindable = unwindable
+        self.corrupt = corrupt
+
+    def mnmemonic_array(self):
+        if self.bytecode_array:
+            return EHABIBytecodeDecoder(self.bytecode_array).mnemonic_array
+        else:
+            return None
+
+    def __repr__(self):
+        return "<EHABIEntry function_offset=0x%x, personality=%d, %sbytecode=%s>" % (
+            self.function_offset,
+            self.personality,
+            "eh_table_offset=0x%x, " % self.eh_table_offset if self.eh_table_offset else "",
+            self.bytecode_array)
+
+
+class CorruptEHABIEntry(EHABIEntry):
+    """ This entry is corrupt. Attribute #corrupt will be True.
+    """
+
+    def __init__(self):
+        super(CorruptEHABIEntry, self).__init__(function_offset=None, personality=None, bytecode_array=None,
+                                                corrupt=True)
+
+    def __repr__(self):
+        return "<CorruptEHABIEntry>"
+
+
+class CannotUnwindEHABIEntry(EHABIEntry):
+    """ This function cannot be unwind. Attribute #unwindable will be False.
+    """
+
+    def __init__(self, function_offset):
+        super(CannotUnwindEHABIEntry, self).__init__(function_offset, personality=None, bytecode_array=None,
+                                                     unwindable=False)
+
+    def __repr__(self):
+        return "<CannotUnwindEHABIEntry function_offset=0x%x>" % self.function_offset
+
+
+class GenericEHABIEntry(EHABIEntry):
+    """ This entry is generic model rather than ARM compact model.Attribute #bytecode_array will be None.
+    """
+
+    def __init__(self, function_offset, personality):
+        super(GenericEHABIEntry, self).__init__(function_offset, personality, bytecode_array=None)
+
+    def __repr__(self):
+        return "<GenericEHABIEntry function_offset=0x%x, personality=0x%x>" % (self.function_offset, self.personality)
+
+
+def arm_expand_prel31(address, place):
+    """
+       address: uint32
+       place: uint32
+       return: uint64
+    """
+    Location = address & 0x7fffffff
+    if Location & 0x04000000:
+        Location |= 0xffffffff80000000
+    return Location + place & 0xffffffffffffffff
```

Comment:
I suspect there's already code in tree for decoding this type of relocation.

I think you should add an entry for `R_ARM_PREL31` to `_RELOCATION_RECIPES_ARM ` in elf/relocation.py, then use...hmm...that might not be as straightforward as I thought.

Anyways, I don't think the local variable `Location` should be capitalized. Also, `offset` would be more appropriate than `place` for an identifier.

Reply:
First, `elf/relocation.py` is too heavy for me. It provides relocation for whole elf file, rather than single entry.

Second, I think `R_ARM_PREL31` and `arm_expand_prel31` are two different concepts. `R_ARM_PREL31` is relocation type in `.rel.ARM.exidx` section.  `arm_expand_prel31` is like an encoding format.

`Location` is my fault, I just copy the code from binutils/readelf.c.